Support ARM instance types

mcrute opened the pull request:
This depends on the `template-updates` in #67 

The magic numbers for console speed, nvme timeout, and block offset all come from the Amazon Linux and Ubuntu AMIs.

We'll want to remove the `install_updated_aarch64_kernel` function once the upstream virt kernel has the ena patches. This should be within the next day or so for edge and in the 3.12 release. I'll push some changes to this to integrate the test profile into the alpine one but wanted to get this out for review as soon as it was cleaned up.

mcrute commented:
Removed the hacks since 3.12 and edge changes are in the repos. Going to try to build 3.12 this weekend.

tomalok commented:
first attempt at adding aarch64 to my custom build...
```
==> Some builds didn't complete successfully and had errors:
--> amazon-ebssurrogate: Error launching source instance: Unsupported: Your requested instance type (a1.medium) is not supported in your requested Availability Zone (us-west-2a). Please retry your request by not specifying an Availability Zone or choosing us-west-2b, usstatus code: 400, request id: 5ec1f6c6-808d-4540-97ab-a4b6b4cde83f
```

mcrute commented:
Is there possibly something that's overriding your AZ preference? I didn't encounter this in my testing.

tomalok commented:
Availability Zone isn't specified in any defaults, I was specifying the subnet to build in, which was in us-west-2a.  Removing that from my profile (as it's no longer in the other profiles) leaves me with...
```
--> amazon-ebssurrogate: Error launching source instance: VPCResourceNotSpecified: The specified instance type can only be used in a VPC. A subnet ID or network interface ID is required to carry out the request.
```

How are you specifying a VPC or subnet?  (I believe those references were removed from the profiles.)

mcrute commented:
That's odd, I removed them from the profiles because they weren't necessary for any of the building I did (even x86_64). Looking at the packer docs it claims that it'll use the default VPC and an available subnet from that VPC if one is setup. I validated that there was a default VPC before I removed them.

I'll clear out my environment and do some testing to see if I can reproduce the issues. I guess we could re-add those if necessary.

mcrute commented:
One of the reasons I removed them is that they can possibly cause breakages when there are no instances of they type requested in a given AZ. But it seems that, for some reason, this is having the exact opposite effect in your test case. Hrm...

tomalok commented:
successful when I specify a subnet in us-west-2b...
```
alpine@ip-192-168-2-41:~$ uname -a
Linux ip-192-168-2-41 5.4.43-1-virt #2-Alpine SMP Thu, 28 May 2020 20:13:48 UTC aarch64 Linux
```

mcrute commented:
Was auto-closed by cleaning up merged branches. Everything is on `master`. But I'll keep an eye on this for feedback from your testing.

tomalok commented:
> One of the reasons I removed them is that they can possibly cause breakages when there are no instances of they type requested in a given AZ. But it seems that, for some reason, this is having the exact opposite effect in your test case. Hrm...

So, my `us-west-2a` may not be the same as your `us-west-2a` -- the logical to physical mappings of AZs is randomized from account to account.

I think in my case, I removed the default VPC some years ago.  I'll try to re-designate or re-create a default.

mcrute commented:
Yeah. My unstated assumption was that you were using the `alpine-amis-user` account, which is where I was doing my testing. If your account is laid out differently it would make sense to import the base profile and override the region specific things for that account.

tomalok commented:
It appears that my account is so old that it doesn't support the automagic creation of default VPCs...  :/  Looking for a way to convert it, I haven't launched anything EC2-classic in ages...

tomalok commented:
Apparently no way to convert my account, so got to play a little with AWS organizations to make a new dev account.  Default VPC worked as it should.

Hopefully the `deploy` portion of `builder.py` will be able to deploy to other accounts as well as other regions?

mcrute commented:
The current implementation uses whatever account your credentials are tied to (or the identity broker, which currently implies a single account). Is that adequate or is there a use-case for specifying cross-account replication?

Also, I'm hoping to ship that release tool by EOW, kinda got caught out by life between the first set of updates and the last.

tomalok commented:
life can be pesky like that, sometimes...  ;)

my new dev/build account isn't my prod account, and i may set up other prod accounts at some point in the future...

the main thing is that i don't want to have to export the new custom AMIs to each account...  I suppose I could keep them in the dev/build account and specify somehow that the other accounts can use them.  If i remember the Packer stuff correctly, `ami_groups` was a way to specify public, private, or (maybe) a list of accounts to grant access?
